homeassistant/components/stream/core.py

```diff
@@ -0,0 +1,146 @@
+"""Provides core stream functionality."""
+import asyncio
+import io
+from typing import List, Any
+
+import attr
+from aiohttp import web
+
+from homeassistant.components.http import HomeAssistantView
+from homeassistant.helpers.event import async_call_later
+
+from .const import DOMAIN, ATTR_STREAMS
+
+
+@attr.s
+class StreamBuffer:
+    """Represent a segment."""
+
+    segment = attr.ib(type=io.BytesIO)
+    output = attr.ib()               # type=av.OutputContainer
+    vstream = attr.ib()              # type=av.VideoStream
+    astream = attr.ib(default=None)  # type=av.AudioStream
+
+
+@attr.s
+class Segment:
+    """Represent a segment."""
+
+    sequence = attr.ib(type=int)
+    segment = attr.ib(type=io.BytesIO)
+    duration = attr.ib(type=float)
+
+
+class StreamOutput:
+    """Represents a stream output."""
+
+    num_segments = 3
+
+    def __init__(self) -> None:
+        """Initialize a stream output."""
+        self.__cursor = None
+        self.__event = asyncio.Event()
+        self.__segments = []
+
+    @property
+    def format(self) -> str:
+        """Return container format."""
+        return None
+
+    @property
+    def audio_codec(self) -> str:
+        """Return desired audio codec."""
+        return None
+
+    @property
+    def video_codec(self) -> str:
+        """Return desired video codec."""
+        return None
+
+    @property
+    def segments(self) -> List[int]:
+        """Return current sequence from segments."""
+        return [s.sequence for s in self.__segments]
+
+    @property
+    def target_duration(self) -> int:
+        """Return the average duration of the segments in seconds."""
+        durations = [s.duration for s in self.__segments]
+        return round(sum(durations) // len(self.__segments)) or 1
+
+    def get_segment(self, sequence: int = None) -> Any:
+        """Retrieve a specific segment, or the whole list."""
+        if not sequence:
+            return self.__segments
+
+        for segment in self.__segments:
+            if segment.sequence == sequence:
+                return segment
+        return None
+
+    async def recv(self) -> Segment:
+        """Wait for and retrieve the latest segment."""
+        last_segment = max(self.segments, default=0)
+        if self.__cursor is None or self.__cursor <= last_segment:
+            await self.__event.wait()
+
+        if not self.__segments:
+            return None
+
+        segment = self.__segments[-1]
+        self.__cursor = segment.sequence
+        return segment
+
+    async def put(self, segment: Segment) -> None:
```

Comment:
`asyncio.event.set()` is not a coroutine, therefore we can remove async and all calls for `put` will be easier.
```suggestion
    def put(self, segment: Segment) -> None:
```

Reply:
I'm not familiar enough with how this works.  Will this cause an issue with how `put` is called in the worker thread?  Currently we are using `asyncio.run_coroutine_threadsafe` there.  Would it be safe to just call `put` directly here?

Reply:
Ah, I remember why I did it this way now: `asyncio.Event` is [not thread-safe](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Event).